File: Recursion/Lecture 14/CodeStudio/Combination Sum III.cpp
void help(int i, int k, int sumTillNow, int n, vector<int> &subSet, vector<vector<int>> &ans)
{
    // base condition
    // if the sumTillNow becomes greater than n, then we don't need to proceed further
    if (sumTillNow > n)
        return;

    // base condition
    // if k becomes equal to zero(0) , then check if the sumTillNow is equal to n or not
    if (k == 0)
    {
        // if the sumTillNow is equal to n, then include the subset into ans[][]
        if (sumTillNow == n)
        {
            ans.push_back(subSet);
        }

        // else we simply return without including the subset
        return;
    }

    // base condition
    // if i becomes equal to 10, then we don't need to proceed any further
    if (i == 10)
        return;

    // pick the ith element
    sumTillNow += i;
    subSet.push_back(i);

    // ask recursion to do the rest of the task
    help(i + 1, k - 1, sumTillNow, n, subSet, ans);

    // backtrack and undo the changes that have been done
    sumTillNow -= i;
    subSet.pop_back();

    // skip ith element
    help(i + 1, k, sumTillNow, n, subSet, ans);
}
vector<vector<int>> combinationSum3(int k, int n)
{
    // Write your code here.
    vector<int> subSet;
    vector<vector<int>> ans;
    help(1, k, 0, n, subSet, ans);
    return ans;
}

/*
    time complexity: O(2^K)
    space complexity: O(K)
*/

File: Recursion/Lecture 14/LeetCode/Combination Sum III.java
/* This question is the continuation of Combination Sum 1 & 2
The only difference is that, instead of giving us an array 
We are asked to print all possible combinations using numbers 1 to 9 (both inclusive)
*/

class Solution
{
    public void help(int i, int k, int sumTillNow, int n, List<Integer> subSet, List<List<Integer>> ans) 
    {
        // If the Sum-Till-Now becomes greater than n, we don't need to proceed further

        if(sumTillNow > n) 
            return ;

        // If k becomes equal to zero(0) , we need to check if the Sum-Till-Now is equal to n or not

        if(k == 0) 
        {
            // If the Sum-Till-Now is equal to n, we include the Subset into our ans

            if(sumTillNow == n) 
            {
                ans.add(new ArrayList(subSet)); 
            }

            // Else we simply return without including the Subset

            return;
        }

        // If i becomes equal to 10, we don't need to proceed any further

        if(i == 10) 
            return ;

        // We include the i-th Number into our Sum-Till-Now & also into our Subset

        sumTillNow += i ;
        subSet.add(i) ;

        // Ask Recursion to do the rest of the task

        help(i + 1, k - 1, sumTillNow, n, subSet, ans) ;

        // Backtrack and undo the change we have done

        sumTillNow -= i ;
        subSet.remove(subSet.size()-1) ;

        // We don't include the i-th Number and ask recursion to do the rest of the Combinations

        help(i + 1, k, sumTillNow, n, subSet, ans); 

    }
    public List<List<Integer>> combinationSum3(int k, int n) 
    {
        List<Integer> subSet = new ArrayList<>();
        
        List<List<Integer>> ans = new ArrayList<>();

        help(1, k, 0, n, subSet, ans) ;

        return ans ;
    }
}

/* 
Time Complexity: O(2^N)
Space Complexity: O(K)
*/
